L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+  // the certificate data in the internal stack.
+  // 
+  // For the parameters in the callback function:
+  // cert_name The name of the certificates being watched.
+  // root_being_watched If the root certificates with the specific name are being
+  // watched.
+  // identity_being-Watched If the identity certificates with the specific name
+  // are being watched.
+  virtual void WatchStatusCallback(std::string cert_name, bool root_being_watched, bool identity_being_watched) = 0;
+
+  // Sets the key materials based on their certificate name.
+  // @param cert_name The name of the certificates being updated.
+  // @param pem_root_certificates The content of root certificates.
+  // @param pem_key_cert_pairs The content of identity key-certificate pairs.
+  void SetKeyMaterials(
+      const std::string& cert_name, absl::optional<std::string> pem_root_certificates,
+      absl::optional<grpc_core::PemKeyCertPairList> pem_key_cert_pairs);
+
+  // Propagates an error encountered in the provider layer to the internal TLS stack.
+  //
+  // @param cert_name The watching certificate name that the caller
+  // wants to notify about when encountering error.
+  // @param root_cert_error The error that the caller encounters when reloading
+  // root certificates.
+  // @param identity_cert_error The error that the caller encounters when
+  // reloading identity certificates.
+  void SetErrorForCert(const std::string& cert_name,
+                       absl::Status root_cert_error,
+                       absl::Status identity_cert_error);
+
+ private:
+  std::unique_ptr<TlsCertificateDistributor> distributor_;
+};
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Provides identity credentials and root certificates.
+*/
+// This represents a potential decoupling of roots and identity chains, with
+// further extension points for something like SPIFFE bundles
+class CertificateProviderInterface {
+ public:
+  virtual ~CertificateProviderInterface() = default;
+  // Provider implementations MUST provide a LoadAndSetCredentialsCallback that
+  // will be called by the internal stack. This will be invoked when a new
+  // certificate name is starting to be used internally, or when a certificate
+  // name is no longer being used internally. When being invoked for a new
+  // certificate, this callback should call SetRootCertificates or
+  // SetIdentityChainAndPrivateKey to do the initial population the certificate
+  // data in the internal stack.
+  // 
+  // For the parameters in the callback function: cert_name The name of the
+  // certificates being watched.  type The type of certificates being watched.
+  virtual void LoadAndSetCredentialsCallback(std::string name, CredentialType type) = 0;
+
+  // Must be called after the constructor.
+  // Does important internal setup steps.
+  virtual void Init() final;
+
+  enum CredentialType {
```

Comment:
From review, explain whether the SPIFFE case falls into the "RootCertificates" bucket or is a separate bucket.